### PR TITLE
feat(devtools): enabled option useAtomsDevtools

### PR DESCRIPTION
**Pull request by Aslemammad**
This PR tries to achieve what we achieved in https://github.com/pmndrs/zustand/pull/880 so we can add an option called `enabled` for better testing/dev env friendly errors/warnings!

This PR is for useAtomsDevtools, so we can make reviewing easier. after this, we'll try achieving this for other dev tools.

### CODE REVIEW

File: src/devtools/useAtomsDevtools.ts [hunk 1]
```diff
@@ -52,11 +52,16 @@ const getDevtoolsState = (atomsSnapshot: AtomsSnapshot) => {
...
 }
 
-export function useAtomsDevtools(name: string, scope?: Scope) {
-  const ScopeContext = getScopeContext(scope)
+interface DevtoolsOptions {
+  scope?: Scope
+  enabled?: boolean
+}
+
+export function useAtomsDevtools(name: string, options?: DevtoolsOptions) {
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:60]: Please make it backward compatible.

```ts
export function useAtomsDevtools(name: string, options?: DevtoolsOptions)
/**
 * @deprecated Please use object options
 */
export function useAtomsDevtools(name: string, options?: Scope)
export function useAtomsDevtools(name: string, options?: DevtoolsOptions | Scope) {
  if (typeof options !== 'undefined' && typeof options !== 'object') {
    console.warn('DEPRECATED...')
    options = { scope: options }
```

Check https://github.com/pmndrs/valtio/pull/400 for reference.

File: src/devtools/useAtomsDevtools.ts [hunk 2]
```diff
@@ -52,11 +52,16 @@ const getDevtoolsState = (atomsSnapshot: AtomsSnapshot) => {
...
+  const ScopeContext = getScopeContext(options?.scope)
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:61]: Maybe we want to destructure in advance, so that code looks cleaner.
```ts
  const { enabled, scope } = options || {}
```

File: src/devtools/useAtomsDevtools.ts [hunk 3]
```diff
@@ -52,11 +52,16 @@ const getDevtoolsState = (atomsSnapshot: AtomsSnapshot) => {
...
   const { s: store, w: versionedWrite } = useContext(ScopeContext)
 
-  if (!store[DEV_SUBSCRIBE_STATE]) {
+  if (!store[DEV_SUBSCRIBE_STATE] && options?.enabled && __DEV__) {
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:64]: I don't think we need this.
Ohhh! we can't enable this in PRD anyway! That was my confusion...

So, it should be something like this. Open for suggestions.
```js
if (enabled === false) {
  return
}
if (!store[DEV_SUBSCRIBE_STATE]) {
  throw new Error('useAtomsDevtools can only be used in dev mode.')
}
```

File: src/devtools/useAtomsDevtools.ts [hunk 4]
```diff
@@ -117,18 +122,21 @@ export function useAtomsDevtools(name: string, scope?: Scope) {
   let extension: typeof window['__REDUX_DEVTOOLS_EXTENSION__']
 
   try {
-    extension = window.__REDUX_DEVTOOLS_EXTENSION__
+    extension =
+      ((options?.enabled ? options.enabled : __DEV__) &&
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:126]: This isn't what we want. It should be `typeof enabled === 'boolean' ? enabled : __DEV__`, to deal with `enabled=false`.


Hmm, it would be much cleaner if we do:
```js
  enabled ?? __DEV__
```
(we should do this in zustand and valtio too.)

File: src/devtools/useAtomsDevtools.ts [hunk 5]
```diff
@@ -117,18 +122,21 @@ export function useAtomsDevtools(name: string, scope?: Scope) {
...
-    if (__DEV__ && typeof window !== 'undefined') {
+    if (__DEV__ && options?.enabled) {
       console.warn('Please install/enable Redux devtools extension')
     }
   }
 
-  if (!store[DEV_SUBSCRIBE_STATE]) {
+  if (!store[DEV_SUBSCRIBE_STATE] && options?.enabled && __DEV__) {
     throw new Error('useAtomsSnapshot can only be used in dev mode.')
   }
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:141]: Why do we have this again? It seems like a leftover. `useAtomsSnapshot` 👀

**Review by dai-shi — CHANGES_REQUESTED** [review]
first review

File: src/devtools/useAtomsDevtools.ts (hunk 4 above)
Comment (Aslemammad) [reply src/devtools/useAtomsDevtools.ts:126]: not putting undefined causes type problems!

File: src/devtools/useAtomsDevtools.ts (hunk 3 above)
Comment (Aslemammad) [reply src/devtools/useAtomsDevtools.ts:64]: This makes problems for us! make tests failing!

File: src/devtools/useAtomsDevtools.ts [hunk 6]
```diff
@@ -52,11 +53,29 @@ const getDevtoolsState = (atomsSnapshot: AtomsSnapshot) => {
...
+) {
+  if (typeof options !== 'undefined' && typeof options !== 'object') {
+    options = { scope: options }
+  }
+  const { enabled, scope } = options || {}
   const ScopeContext = getScopeContext(scope)
   const { s: store, w: versionedWrite } = useContext(ScopeContext)
 
-  if (!store[DEV_SUBSCRIBE_STATE]) {
+  if (!store[DEV_SUBSCRIBE_STATE] && enabled && __DEV__) {
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:78]: I still want the early return.
Which test fails? I mean it should only fail if `enabled === false`, right? It should be a new test.

```suggestion
  if (enabled === false) {
    return
  }
  if (!store[DEV_SUBSCRIBE_STATE]) {
```

**Review by dai-shi — CHANGES_REQUESTED** [review]
second one!

File: src/devtools/useAtomsDevtools.ts (hunk 6 above)
Comment (dai-shi) [reply src/devtools/useAtomsDevtools.ts]: I completely forgot about this being a hook. We can't early return. That's really my bad. So sorry to have your time wasted. Added a commit: dc485b7

File: src/devtools/useAtomsDevtools.ts [hunk 7]
```diff
@@ -114,83 +152,68 @@ export function useAtomsDevtools(name: string, scope?: Scope) {
...
-    throw new Error('useAtomsSnapshot can only be used in dev mode.')
-  }
-
   const isTimeTraveling = useRef(false)
   const isRecording = useRef(true)
   const devtools = useRef<
-    ReturnType<NonNullable<typeof extension>['connect']> & {
+    ReturnType<
+      NonNullable<typeof window['__REDUX_DEVTOOLS_EXTENSION__']>['connect']
+    > & {
```
Comment (dai-shi) [inline src/devtools/useAtomsDevtools.ts:160]: Is this change necessary?

**Review by dai-shi — APPROVED** [review]
